notify/silences.go

```diff
@@ -82,35 +83,58 @@ func (am *GrafanaAlertmanager) GetSilence(silenceID string) (GettableSilence, er
 
 // CreateSilence persists the provided silence and returns the silence ID if successful.
 func (am *GrafanaAlertmanager) CreateSilence(ps *PostableSilence) (string, error) {
+	sil, err := am.marshalSilence(ps)
+	if err != nil {
+		return "", err
+	}
+
+	silenceID, err := am.silences.Set(sil)
+	if err != nil {
+		level.Error(am.logger).Log("msg", "unable to save silence", "err", err)
+		return "", fmt.Errorf("unable to save silence: %s: %w", err.Error(), ErrCreateSilenceBadPayload)
+	}
+
+	return silenceID, nil
+}
+
+// UpsertSilence allows for the creation of a silence with a pre-set ID.
+func (am *GrafanaAlertmanager) UpsertSilence(ps *PostableSilence) (string, error) {
+	sil, err := am.marshalSilence(ps)
+	if err != nil {
+		return "", err
+	}
+
+	silenceID, err := am.silences.Upsert(sil)
+	if err != nil {
+		level.Error(am.logger).Log("msg", "unable to upsert silence", "err", err)
+		return "", fmt.Errorf("unable to upsert silence: %s: %w", err.Error(), ErrCreateSilenceBadPayload)
+	}
+
+	return silenceID, nil
+}
+
+// marshalSilence converts a PostableSilence to a Silence protobuf and validates it.
+func (am *GrafanaAlertmanager) marshalSilence(ps *PostableSilence) (*silencepb.Silence, error) {
 	sil, err := v2.PostableSilenceToProto(ps)
 	if err != nil {
 		level.Error(am.logger).Log("msg", "marshaling to protobuf failed", "err", err)
-		return "", fmt.Errorf("%s: failed to convert API silence to internal silence: %w",
```

Comment:
nit: you could inline `PostableSilenceToProto` and then just call this function validate silence - feels more natural than calling a function `marshalSilence` to convert to protobuf (typically these are suffixed with `ToProto`) *AND* valiting it.